### ISSUE TITLE
Can libvmi get page error?

### ISSUE THREAD

**Issue by Badangel** [libvmi/libvmi#570]
I want to catch syscall by use wrong address. When it happen page error, vm get into vmm, whether libvmi can get it. How to set it and use which funtion. I can use software int to catch syscall. The event of libvmi  use SETUP_INTERRUPT_EVENT. It is very slow and some syscall can't catch. Wish your reply! Thanks.

**Comment by Wenzel** [comment]
@Badangel why don't you use a mem_event to catch the syscalls ?
do you have the addresses of all the syscalls ?
for example you can dump the SSDT on Windows.

**Comment by Badangel** [comment]
@Wenzel I have all addresses of syscalls, but I use mem_event can't catch it in ubuntu
such as, I know sys_open address is 0xffffffff8120d9c0 
I use this code to register
```
   addr_t sysopen_addr = 0xffffffff8120d9c0;
   addr_pa=vmi_translate_kv2p(vmi, sysopen_addr );
    gfn = addr_pa>> 12;
    SETUP_MEM_EVENT(&mm_event, sysopen_addr, VMI_MEMACCESS_X, mm_callback, 0);
    vmi_register_event(vmi, &mm_event);
```
I read rax in mm_callback. It should be open sysunm 2. But it is wrong and the VM suspend and stuck.
```
event_response_t mm_callback(vmi_instance_t vmi, vmi_event_t *event)
{
    vmi_get_vcpureg(vmi, &cr3, CR3, 0);
    current_pid=vmi_dtb_to_pid(vmi, cr3);
    reg_t rax;
    vmi_get_vcpureg(vmi, &rax, RAX, 0);
    printf("Memevent: \tPID %u. RAX %ld:\n", current_pid, rax);
    return 0;
}
```
Do you know why? And how to use this funtion, Can you give me a sample?
Thank you very much!

**Comment by Wenzel** [comment]
@Badangel do not use get_vcpureg inside a callback, since it queries the register state which have not been updated yet by the new values from the event.

Instead, use `event->x86_regs->XXX`:
example: https://github.com/Wenzel/radare2-extras/blob/vmi/vmi/debug_vmi.c#L61

**Comment by Wenzel** [comment]
@Badangel another thing that i forgot is that with memory events, you need to check if you are at the `rip` you were looking for.

a memory event is triggered when the code reaches a certain page.
You need to check if rip is the rip of your syscall address, and otherwise, just emulate the instruction.
here i'm returning `VMI_EVENT_RESPONSE_EMULATE` because RIP is not on my breakpoint address:
https://github.com/Wenzel/radare2-extras/blob/vmi/vmi/debug_vmi.c#L62

**Comment by Badangel** [comment]
@Wenzel You mean that mem_event only can catch which the page access, and I should make sure the event->x86_regs->rip is the address which I want to catch. And use event->x86_regs->rax to get RAX register value. 
I will try it, thanks!

**Comment by Badangel** [comment]
@Wenzel  I use mem_event to catch syscall. But the result is not very good. 
I catch all syscalls by writing int 3, when I open the firefox need 10s. 
But when I use mem_event only catch sys_open, the time to open firefox need 22s. 
The normal open time only 4s.
I don't know whether it's  normal. And is there a faster way to catch syscall?
No matter which way I use, when I want to get syscall parameter, sometime it will have a error that say the address out of max address. My system is 64bit.
such as sys_open, the first parameter is path of file in rdi.
I get parameter code:
```
addr_t rdifilenameaddr2 = event->x86_regs->rdi;
char *rdifilename2 = vmi_read_str_va(vmi, rdifilenameaddr2, current_pid);
```
error:
> VMI_ERROR: --requesting PA [0x76f646e698000] beyond max physical address [0xff000000]
> VMI_ERROR: 	paddr: 76f646e697000, length 1000, vmi->max_physical_address ff000000
> VMI_ERROR: create_new_entry failed
Look forward to your reply.

**Comment by tklengyel** [comment]
@Badangel if you have trouble writing yout own syscall tracer with LibVMI, try DRAKVUF, it supports Linux and doing what you are trying to do is probably simpler with that then going straight with LibVMI.

**Comment by Badangel** [comment]
@tklengyel Do you know why it have VMI_ERROR: out of max PA?

**Comment by tklengyel** [comment]
Nope, the address it tries to read doesn't look like a valid address at all. Could perhaps happen if the pagetable through which that address was obtained via was in the middle of being setup so it would still have uninitialized memory?
